### PR TITLE
Fixes an issue on layer editor when trying to save a root layer directly (by right clicking on the layer and Save As.), where an invalid extension could cause a crash. The fallback is .usd.

### PR DESCRIPTION
**Pull request by AramAzhari-adsk**
Fixes an issue on layer editor when trying to save a root layer directly (by right clicking on the layer and Save As.), where an invalid extension could cause a crash. The fallback is .usd.

It also blocks the path that will result in the crash for any other scenario where the file cannot be opened after it was exported. The user is then displayed a message dialog indicating the layer saving was not successful.

Similar PR: https://github.com/Autodesk/maya-usd/pull/2845

### CODE REVIEW

File: lib/usd/ui/layerEditor/layerTreeItem.cpp [hunk 1]
```diff
@@ -363,22 +364,34 @@ void LayerTreeItem::saveAnonymousLayer()
...
+
+        std::string        filePath(fileName);
+        const std::string& extension = SdfFileFormat::GetFileExtension(filePath);
+        const std::string  defaultExt(UsdMayaTranslatorTokens->UsdFileExtensionDefault.GetText());
+        const std::string  usdCrateExt(UsdMayaTranslatorTokens->UsdFileExtensionCrate.GetText());
+        const std::string  usdASCIIExt(UsdMayaTranslatorTokens->UsdFileExtensionASCII.GetText());
+        if (extension != defaultExt && extension != usdCrateExt && extension != usdASCIIExt) {
+            filePath.append(".");
+            filePath.append(defaultExt.c_str());
+        }
```
Comment (pierrebai-adsk) [inline lib/usd/ui/layerEditor/layerTreeItem.cpp:376]: If this is the same code as in the other review, maybe create a function, something like ensureUSDFileExtension() or similar.

File: lib/usd/ui/layerEditor/layerTreeItem.cpp [hunk 2]
```diff
@@ -363,22 +364,34 @@ void LayerTreeItem::saveAnonymousLayer()
 
     std::string fileName;
     if (sessionState->saveLayerUI(nullptr, &fileName)) {
+
+        std::string        filePath(fileName);
```
Comment (pierrebai-adsk) [inline lib/usd/ui/layerEditor/layerTreeItem.cpp:368]: Not sure why we need to create a new variable here. I think we could process fileName directly.

**Review by pierrebai-adsk — CHANGES_REQUESTED** [review]
Almost there, just a few improvements that I think would be worthwhile.

File: lib/usd/ui/layerEditor/layerTreeItem.cpp [hunk 3]
```diff
@@ -363,22 +364,34 @@ void LayerTreeItem::saveAnonymousLayer()
...
         std::string   formatTag = MayaUsd::utils::usdFormatArgOption();
-        if (saveSubLayer(dialogTitle, parentLayerItem(), layer(), fileName, formatTag)) {
-            printf("USD Layer written to %s\n", fileName.c_str());
+        if (saveSubLayer(dialogTitle, parentLayerItem(), layer(), filePath, formatTag)) {
+            printf("USD Layer written to %s\n", filePath.c_str());
 
             // now replace the layer in the parent
             if (isRootLayer()) {
+                sessionState->rootLayerPathChanged(filePath);
                 sessionState->rootLayerPathChanged(
```
[the remark this answers is not in the document]
Comment (AramAzhari-adsk) [reply lib/usd/ui/layerEditor/layerTreeItem.cpp:387]: Good catch! Result of a poor merge.

File: lib/usd/ui/layerEditor/layerTreeItem.cpp (hunk 1 above)
Comment (AramAzhari-adsk) [reply lib/usd/ui/layerEditor/layerTreeItem.cpp]: done

File: lib/usd/ui/layerEditor/layerTreeItem.cpp (hunk 2 above)
Comment (AramAzhari-adsk) [reply lib/usd/ui/layerEditor/layerTreeItem.cpp]: Removed the extra variable.

**Review by pierrebai-adsk — DISMISSED** [review]

File: lib/usd/ui/layerEditor/layerTreeItem.cpp [hunk 4]
```diff
@@ -10,6 +10,7 @@
 #include "warningDialogs.h"
 
 #include <mayaUsd/base/tokens.h>
+#include <mayaUsd/fileio/jobs/jobArgs.h>
```
Comment (vlasovi) [inline lib/usd/ui/layerEditor/layerTreeItem.cpp:13]: Is this include necessary?

**Review by vlasovi — DISMISSED** [review]
I think including "mayaUsd/fileio/jobs/jobArgs.h" is not necessary. Otherwise looks good

File: lib/usd/ui/layerEditor/layerTreeItem.cpp (hunk 4 above)
Comment (AramAzhari-adsk) [reply lib/usd/ui/layerEditor/layerTreeItem.cpp:13]: This is no longer necessary as the function now lives in utilSerialization. I'll remove it.

**Review by vlasovi — APPROVED** [review]

**Review by vlasovi — APPROVED** [review]

**Review by seando-adsk — APPROVED** [review]

**Pull request merged**